### DISCUSSION TITLE
Simple interface for PandaPower

### DISCUSSION THREAD

**Discussion by jinningwang** [CURENT/andes#265] [Ideas]
[03/02/2023]
***Information related to this interface***
We are now developing a new dispatch software AMS that aims to do co-simulation with ANDES. It is still under active development and is expected to be released this year. Drop me a message if you have any thoughts about dispatch-dynamic co-simulation.
Welcome to check out the [AMS code repository](https://github.com/jinningwang/ams). Stay tuned! 

---
[02/06/2022]
Hi Hantao,

I'm doing some study with OPF. I think that maybe we can develop a simple interface between PandaPower and ANDES I/O file, so the dispatch results can be assigned to dynamic files.
On my horizon, the workload should be fine, I'm wondering do you think it is doable and useful?

Regards,
Jinning

**Comment by cuihantao** [comment]
That will be a useful feature. Any initial thoughts on the design?

**Reply by jinningwang** [reply]
The preliminary idea is to define a method (or class) ``interface``. The following aspects should be considered:

1) ``interface`` should include the necessary consistency checks.
2) ``interface`` function ``ans2pdp``: read an ANDES system model and output the data required for PandaPower.
3) ``interface`` function ``pdp2ans``: read the OPF results from PandaPower and write the data into an ADNES system model.

Issues:
1) I am not sure which definition will be better, a class or a method? (a) method: will the two functions be overwhelming for it? (b) class: a better choice to hold many such contents, but then where should we build this class?
2) I am using PandaPower now, and Xin may provide some help. An example code can be really helpful.
3) The detailed code structure is not prepared yet, but I can try to build it step by step.

What do you think?

Regards,
Jinning

**Reply by jinningwang** [reply]
Another helpful thing is that, since Qiwei has already developed AMS, there should already be a useful data converter? Maybe we can get some readily experience from him.

Regards,
Jinning

**Comment by cuihantao** [comment]
> interface function ans2pdp: read an ANDES system model and output the data required for PandaPower.
> interface function pdp2ans: read the OPF results from PandaPower and write the data into an ADNES system model.

If you write them from the andes side, then can be `pp_system = to_pandapower(system)` and `system = from_pandapower(pp_system)`.

> I am not sure which definition will be better, a class or a method? (a) method: will the two functions be overwhelming for it? (b) class: a better choice to hold many such contents, but then where should we build this class?

It doesn't matter that much. Classes are useful when you need to store and reuse internal states. For this case, a function calling a list of subfunctions be straightforward. See the parser for PSS/E raw file.

> I am using PandaPower now, and Xin may provide some help. An example code can be really helpful.
You can check with Xin. I know some interface was written.

> Another helpful thing is that, since Qiwei has already developed AMS, there should already be a useful data converter? Maybe we can get some readily experience from him.
You can also check with Qiwei. Not sure about the status here.

**Reply by jinningwang** [reply]
Thanks for your reply, now it looks like the workload for the interface is acceptable.

**Comment by MaKla89** [comment]
Greetings @jinningwang ! I am currently trying to do the same (though manually, for now). How did you map the PP-OPF generator-settings to the ANDES generators (or turbine govenors)? :)  I am able to use TGOV1N and set P as pref0, but I have no clue how to set either ref-voltage from OPF-results or Q-setpoint from OPF-results.

**Reply by jinningwang** [reply]
Hi,

Dispatching the voltage setting points is kind of tricky.
Now, the voltage setting points can be done by initially setting the voltage of the StaticGen, and then once the dynamic simulation is initialized, the voltage of the dynamic generator will be set as the same as the StaicGen.
However, changing the voltage setting points in a dynamic simulation is not considered right now, because ``vref`` of an Exciter is not exactly the terminal voltage of a generator.
You can gradually change ``vref`` of an exciter to obtain the desired terminal voltage, although it is not designed to do so.

@cuihantao 
Hi Hantao,

Do you think it is reasonable to use an "external PI controller" which tunes the generator voltage by tuning the ``vref`` of its exciter?

Regards,
Jinning

**Reply by MaKla89** [reply]
Hi Jinning,

thanks A LOT for answering (I know this post has already been a bit older ;) ). And I agree, changing something like `vref` does neither seem very intuitive nor too easy to implement. Two questions to you at this stage:

1. How did YOU implement the mapping from OPF results back to the dynamic simulation? (or did you not change the dynamic-simulation generator settings after initialisation?)

2. Do you think changing P + Q during the dynamic simulation would be more reasonable than P + V , if I want to apply the OPF-results to a running simmulation?

Thanks again! :)

**Reply by jinningwang** [reply]
By the way, an example can be helpful to you. https://docs.andes.app/en/latest/examples/pandapower.html#generator-dispatch-based-on-opf-from-pandapower
We can change ``StaicGen.v0`` to change the voltage setting points of a Static Generator before initializing the dynamic simulation.

Regards,
Jinning

**Reply by MaKla89** [reply]
> By the way, ...

Absolutely correct :) Sadly I'd need to apply the OPF-setpoints to the generators "on the fly" and cannot rely on re-initialisation in my specific case. If nothing else, I'll try your suggestion of changing `vref` of the exciter, though I am not sure that I have seen this setpoint yet (then again not all of my models do even use exciters, maybe I just missed this bit :) )

Thanks again!

**Reply by jinningwang** [reply]
> 

1. You can check the example I posted. If necessary, you can check the source code of ``runopp_map``.
2. In the dynamic simulation, changing Q can be done by switching the reactive power compensator. It should be easier than changing the Q of a generator. However, the trouble is that you need to carefully calculate how much Q you need for a given voltage.

Regards,
Jinning

**Reply by MaKla89** [reply]
Excellent, thank you! I am working with that example right now, but could only find the pref0 (which is not bad as a nice confirmation that I am on the right track :) ). Nevermind, you helped me a lot! The compensator idea is also nice, but, as you pointed out, quite tricky to calculate. I'll consider it a last resort ;)

Thank you!

**Reply by cuihantao** [reply]
@jinningwang Of course a PI controller can achieve desired reference voltage. The output of the PI can go to `vi`, which by convention is the summation block for `vref` and other input signals. It's not very compelling to me to add such a support for all exciters.

@MaKla89 I think setting target voltage is more practical than setting the desired Q. Reactive power compensation is a measure to keep voltages within the range.

**Reply by jinningwang** [reply]
@cuihantao Many thanks for your reply! We can do this when necessary.

@MaKla89 I agree with Dr. Cui's opinion, it is more reasonable to control the voltage rather than using the reactive power compensation.

**Comment by MaKla89** [comment]
Greetings! Yes, me again, I'm sorry :(

I did check the great [example](https://docs.andes.app/en/stable/examples/pandapower.html) for this interface, as was suggested previously. I tried to copy-paste the example line by line.  With an exception of having to load the plotter, it almost worked out of the box (great!). The only problem:  setting the 'pref0' values does not have any impact on the TDS. I can see that the pref0 values ARE indeed changed via 

`ssa.TurbineGov.set(src='pref0', idx=ssa_gov_idx, attr='v', value=ssp_res['p'][~ssp_res['gov_idx'].isna()])`

but the final plot after continuing TDS to 50 looks like no change in Pe at any time, while the last plot in the example shows the nice workings of the turbine governor. I assume it is something very minor, but as I copy-pasted everything from the example, I am a bit clueless right now. Is there any step (like converting generator types, etc.) that I need to do in order the make the example work?

Sorry for the inconvenience!

**Reply by cuihantao** [reply]
You are really welcome! Thank you for taking the time to provide this valuable feedback!

I ran this example on my machine with the latest version (1.8.5). It gave me the same result as shown in the documentation. 

First, please upgrade the tool if you are not using the latest version. You may download the Jupyter Notebook from here: https://github.com/cuihantao/andes/tree/master/docs/source/examples. See if that fixes the problem.

**Reply by MaKla89** [reply]
Alright, just to own up to my own mistakes (and maybe help someone else in the same situation in the future):

If you exactly copy-paste the code there is ONE (assumed to be minor) problem in line 'ssa.TDS.plt.plot(ssa.GENROU.Pe)'.

This command does not work as the plotter (at least in my test) needs to be loaded first. I did so by adding 'ssa.TDS.load_plotter()'. When 'ssa.TDS.plt.plot(ssa.GENROU.Pe)' is called again at the very end of the example, the plotter needs to be loaded AGAIN, otherwise the plotter still has the data from when it was loaded the first time (d'uh!). This is why it seemed as if the simulation did not account for the OPF-result-based changes. But in reality the TDS was fine, just the plotter was still using old data....

TL;DR: Reload the plotter everytime you want to plot new results!

Recommendation for the example: Maybe add the plotter-loading to the example? :)

**Reply by cuihantao** [reply]
Hello, thank you for taking the time to figure out the issue. That is really helpful.

I was not able to replicate the issue on my end. It may have to do with how Jupyter and matplotlibt is installed. 

In the very first example, loading the plotter is mentioned: https://docs.andes.app/en/latest/_examples/ex1.html#export-and-plot-results. I decided not to duplicate it in subsequent examples. Hope this helps.

**Comment by MaKla89** [comment]
Now that the example is working nicely, I'd have the next question: Would it also be possible to have TDS and PP-OPF run alternating?

Case description: I want to FIRST check the impact of a disturbance on the dynamic behaviour of the power system, THEN I want to run an OPF to see a new set of setpoints for some (all) generators. Then I want to adapt the TDS towards the OPF-setpoints and continue TDS.
MOST of these steps are one by one what has been shown in the Pandapower-Example, I know. But whenever I `to_pandapower(sys)` with `sys` being an ANDES-system post TDS, I get the error `UserWarning: No reference bus is available. Either add an ext_grid or a gen with slack=True` . Are slack-buses / slack-generators converted irrevertably at the start of a TDS or something like that? :)

**Reply by jinningwang** [reply]
Hello,

The pandapower interface does not include standard iterative two-direction conversion right now.
However, you can use the same pandapower system rather than build a new one every time. A straightforward way can be changing the set points in pandapower.

Two clarifications:
1. the pandapower interface does not track the post-TDS system (GENROU, GENCLS, etc.) but only the static power flow system (Bus, PV, PQ, Slack, Shunt). 
2. the pandapower interface convert the ``Slack`` in andes to ``Generator`` in pandapower rather than ``ext_gen``.
 
I'll include these points in future documentation.

Here is a tip that pays attention to the unit conversion because pandapower uses the nominal value, whereas ANDES uses the per-unit value.

**Reply by MaKla89** [reply]
Alright, I dived a bit deeper into your interface to get a better understanding for various things that were unexplainable to me before (like "at what point are all generators set to "NOT in_service", which seems to happen during the TDS and renders mapping the resulting system to Pandapower / pyPower challenging :D).

At this point I partially consider to either modify / extend your interface so that it can also map a post-TDS-andes system to a new PP-system (in order to run OPF). On the coding-side it should be somewhat within my reach (I hope), even if the result might turn out categorically "Quick'n'dirty". For this I might need some initial brainstorming-support regarding "what states / variables would need to be mapped from ANDES (post-TDS) to Pandapower and where in ANDES can I find them?". For static cases (or initial Pre-TDS cases) this seemed doable, but for post-TDS states I lost track on where to find the correct "measurements" to send to Pandapower.

Would someone of you be willing to support me with this high-level task? :)

**Reply by cuihantao** [reply]
It might also be in @jinningwang's interest to map an ANDES system at any point of a TDS to pypower/pandapower.

I will also be available for quick discussions if you need information regarding the data structures.

**Reply by jinningwang** [reply]
Hi MaKla89,

Generator status is controlled by variable ``u``, and all variables including the generator's online status can be retrieved. More details of working with data can be found at [Example2](https://docs.andes.app/en/latest/_examples/ex2.html). The variables you need depend on the work you want to do, and I don't have further comment on this for now.
To my understanding, both creating a new PP system and modifying the previous one can work, and they don't make much difference to me.

Regards,
Jinning

**Reply by MaKla89** [reply]
Perfect, thanks to both of you! I'll discuss the exact requirements internally and come back to you if I have questions :) 

As for the difference between modifying an existing PP-network and creating a new one: As topological changes during the simulated timefrime might theoretically come up, I thought that simply creating a new PP-network might be more straightforward and cleaner.

Just as information: Right now I intend to have OPF-constraints vary depending on whether freuency-deviations are observed or not. Of course the frequency-part is NOT part of the OPF, but it will determine its constraints. (No replied required ;) )

Thanks!
